src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py

```diff
@@ -15028,3 +15028,103 @@ def stack(
             return qc.dropna(axis=0, how="any", thresh=None)
         else:
             return qc
+
+    def unstack(
+        self,
+        level: Union[int, str, list] = -1,
+        fill_value: Optional[Union[int, str, dict]] = None,
+        sort: bool = True,
+        is_series_output: bool = False,
+    ) -> "SnowflakeQueryCompiler":
+        """
+        Pivot a level of the (necessarily hierarchical) index labels.
+
+        Returns a DataFrame having a new level of column labels whose
+        inner-most level consists of the pivoted index labels.
+
+        If the index is not a MultiIndex, the output will be a Series
+        (the analogue of stack when the columns are not a MultiIndex).
+
+        Parameters
+        ----------
+        level : int, str, list, default -1
+            Level(s) of index to unstack, can pass level name.
+
+        fillna : int, str, dict, optional
+            Replace NaN with this value if the unstack produces missing values.
+
+        sort : bool, default True
+            Sort the level(s) in the resulting MultiIndex columns.
+
+        is_series_output : bool, default False
+            Whether the output is a Series, used by `fillna`
+        """
+        level = level if is_list_like(level) else [level]
+        if not sort:
+            ErrorMessage.not_implemented(
+                "Snowpark pandas DataFrame/Series.unstack does not yet support the `sort` parameter"
+            )
+
+        # Check to see if we have a MultiIndex, if we do, make sure we remove
+        # the appropriate level(s), and we pivot accordingly.
+        index_cols = [f"level_{i}" for i in range(self.index.nlevels)]
+        if len(index_cols) > 1:
+            pivot_cols = [index_cols[lev] for lev in level]  # type: ignore
+            index_cols = [
+                index_cols[i]
+                for i in range(len(index_cols))
+                if index_cols[i] not in pivot_cols
+            ]
+            vals = [
+                c for c in self.columns if c not in index_cols and c not in pivot_cols
+            ]
+            qc = (
+                self.reset_index()
+                .pivot_table(
+                    columns=pivot_cols,
+                    index=index_cols,
+                    values=vals,
+                    aggfunc="min",
+                    fill_value=None,
+                    margins=False,
+                    dropna=True,
+                    margins_name="All",
+                    observed=False,
+                    sort=sort,
+                )
+                .set_index_names([None])
+            )
+
+            from snowflake.snowpark.modin.pandas.dataframe import DataFrame
+
+            qc = (
+                DataFrame(query_compiler=qc)
+                .rename_axis([None] * len(qc.columns), axis=1)
+                ._query_compiler
+            )
```

Comment:
Seems like the `rename_axis` implementation is handled by the frontend layer and not pushed down to the query compiler -- couldn't find a better way of handling this. Wonder if folks have other ideas -- I can add a comment if that's helpful.

Reply:
Typically, once we come to backend, we try not to go back to the frontend object. Here seems we just want to reset the column index name field to all None, you can simply copy the internal frame with data_column_index_names field set to all None at this moment, and we should implement rename_axis correctly with Snowpark pandas

Reply:
Given we are interacting with the query compiler the whole way through, creating an InternalFrame that is a copy of all the attributes since it's immutable and just setting `data_column_index_names` to [None] and recreating the query compiler seems like a less clean implementation to me. I added a comment around what this is doing.